<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Simulator for genotype likelihoods. — simgl • hwep</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Simulator for genotype likelihoods. — simgl"><meta property="og:description" content="Uses the updog R package for simulating read counts and generating
genotype log-likelihoods."><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">hwep</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/introduction.html">Introduction to the `hwep` package</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/dcgerard/hwep/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Simulator for genotype likelihoods.</h1>
    <small class="dont-index">Source: <a href="https://github.com/dcgerard/hwep/blob/HEAD/R/bayes.R" class="external-link"><code>R/bayes.R</code></a></small>
    <div class="hidden name"><code>simgl.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Uses the updog R package for simulating read counts and generating
genotype log-likelihoods.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">simgl</span><span class="op">(</span></span>
<span>  <span class="va">nvec</span>,</span>
<span>  rdepth <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  od <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  bias <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  seq <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  ret <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gl"</span>, <span class="st">"gp"</span>, <span class="st">"all"</span><span class="op">)</span>,</span>
<span>  est <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>nvec</dt>
<dd><p>The genotype counts. <code>nvec[k]</code> contains the number of
individuals with genotype <code>k-1</code>.</p></dd>


<dt>rdepth</dt>
<dd><p>The read depth. Lower means more uncertain.</p></dd>


<dt>od</dt>
<dd><p>The overdispersion parameter. Higher means more uncertain.</p></dd>


<dt>bias</dt>
<dd><p>The allele bias parameter. Further from 1 means more bias.
Must greater than 0.</p></dd>


<dt>seq</dt>
<dd><p>The sequencing error rate. Higher means more uncertain.</p></dd>


<dt>ret</dt>
<dd><p>The return type. Should we just return the genotype
likelihoods (<code>"gl"</code>), just the genotype posteriors
(<code>"gp"</code>), or the entire updog output (<code>"all"</code>)</p></dd>


<dt>est</dt>
<dd><p>A logical. Estimate the updog likelihood parameters while
genotype (<code>TRUE</code>) or fix them at the true values (<code>FALSE</code>)?
More realistic simulations would set this to <code>TRUE</code>, but it makes
the method much slower.</p></dd>


<dt>...</dt>
<dd><p>Additional arguments to pass to
<code><a href="https://rdrr.io/pkg/updog/man/flexdog_full.html" class="external-link">flexdog_full</a>()</code>.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>By default, a matrix. The genotype (natural) log likelihoods.
The rows index the individuals and the columns index the dosage. So
<code>gl[i,j]</code> is the genotype log-likelihood for individual i
at dosage j - 1.</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>David Gerard</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">simgl</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span>, model <span class="op">=</span> <span class="st">"norm"</span>, est <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             [,1]      [,2]      [,3]       [,4]      [,5]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1,]  -0.0527924 -4.370039 -9.577024 -17.181178 -43.85351</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2,] -17.2757020 -1.821025 -1.731916  -4.231294 -19.60222</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [3,]  -6.2869622 -1.796049 -4.883182 -10.467745 -32.96015</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [4,] -13.4332263 -1.434198 -2.403487  -5.939004 -23.82247</span>
<span class="r-in"><span><span class="fu">simgl</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span>, model <span class="op">=</span> <span class="st">"norm"</span>, est <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              [,1]      [,2]      [,3]       [,4]      [,5]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1,]  -0.09618742 -2.798880 -6.516204 -12.512266 -31.31682</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2,]  -4.94789634 -1.315725 -3.006407  -7.128083 -22.82756</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [3,]  -4.94789634 -1.315725 -3.006407  -7.128083 -22.82756</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [4,] -10.22692932 -1.899873 -1.614191  -3.834738 -16.08930</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by David Gerard, NSF DBI 2132247.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

